tool_quality_checks/correct_import_file.py

```diff
@@ -0,0 +1,244 @@
+# -*- coding: utf-8 -*-
+"""#
+Created on Tue Aug 13 08:14:18 2019
+@author: chris.kerklaan - N&S
+"""
+
+# system imports
+
+# Third-party imports
+import ogr
+import osr
+import logging
+
+# Globals
+DRIVER_OGR_MEM = ogr.GetDriverByName("Memory")
+DRIVER_OGR_SHP = ogr.GetDriverByName("ESRI Shapefile")
+MEM_NUM = 0
+
+# Exceptions
+logger = logging.getLogger(__name__)
+ogr.UseExceptions()
+
+
+def create_mem_ds():
+    """ Creating an ogr datasource in memory"""
+    global MEM_NUM
+    mem_datasource = DRIVER_OGR_MEM.CreateDataSource("mem{}".format(MEM_NUM))
+    MEM_NUM = MEM_NUM + 1
+    return mem_datasource
+
+
+def create_geom_transform(in_spatial_ref, out_epsg):
+    """ Creating a coordinate transformation based on two reference systems """
+    out_spatial_ref = osr.SpatialReference()
+    out_spatial_ref.ImportFromEPSG(out_epsg)
+    coordTrans = osr.CoordinateTransformation(in_spatial_ref, out_spatial_ref)
+    return coordTrans
+
+
+def multipoly2poly(in_layer, out_layer):
+
+    lost_features = []
+    layer_defn = in_layer.GetLayerDefn()
+    field_names = []
+    for n in range(layer_defn.GetFieldCount()):
+        field_names.append(layer_defn.GetFieldDefn(n).name)
+
+    for count, in_feat in enumerate(in_layer):
+        content = {}
+        for field_name in field_names:
+            content[field_name] = in_feat.GetField(field_name)
+
+        geom = in_feat.GetGeometryRef()
+        if geom == None:
+            logger.warning("FID {} has no geometry.".format(count))
+            lost_features.append(in_feat.GetFID())
+            continue
+
+        if (
+            geom.GetGeometryName() == "MULTIPOLYGON"
+            or geom.GetGeometryName() == "MULTILINESTRING"
+        ):
+            for geom_part in geom:
+                addPolygon(geom_part.ExportToWkb(), content, out_layer)
+        else:
+            addPolygon(geom.ExportToWkb(), content, out_layer)
+
+    return lost_features
+
+
+def addPolygon(simple_polygon, content, out_lyr):
+    featureDefn = out_lyr.GetLayerDefn()
+
+    polygon = ogr.CreateGeometryFromWkb(simple_polygon)
+    out_feat = ogr.Feature(featureDefn)
+    out_feat.SetGeometry(polygon)
+
+    for key, value in content.items():
+        out_feat.SetField(key, value)
+
+    out_lyr.CreateFeature(out_feat)
+
+
+def fix_geometry(geometry):
+    geom_name = geometry.GetGeometryName()
+    # check pointcount if linestring
+    if "LINESTRING" in geom_name:
+        if geometry.GetPointCount() == 1:
+            logger.warning("Geometry point count of linestring = 1")
+            return geometry, False
+        else:
+            pass
+
+    # solve self intersections
+    if "POLYGON" in geom_name:
+        try:
+            geometry = geometry.Buffer(0)
+        except Exception as e:
+            # RuntimeError: IllegalArgumentException: Points of LinearRing do not form a closed linestring
+            logger.warning(e)
+            return geometry, False
```

Comment:
Werkt het goed/duidelijk genoeg om "False" terug te geven? Dan moet je er daarna weer op checken. Daar heeft python op zich juist exceptions voor.

Je verwacht van fix_geometry dat je de gefixte geometry terug krijgt. Maar je kan dus ook de oorspronkelijke terugkrijgen met "False" om aan te geven dat het niet klopt?

Misschien klopt het wel, hoor.

Reply:
ik snap wat je bedoeld, heb de naamgeving nu aangepast, omdat ik even geen handigere oplossing wist hoe dit af te vangen.

Reply:
Verschil is wel met correct_vectory_layer, dat we hier proberen om het script toch door te laten lopen met terugkoppeling dat er iets is misgegaan.

Chris verzameld deze en geeft dat vervolgens in één keer terug, in plaats van per object